client/DESIGN.md

```diff
@@ -0,0 +1,380 @@
+
+- Parameters: 
+    1. server: string (required)
+    2. port: string (port=80)
+    3. max\_age: int (max\_age=None)
+    4. no\_cache: bool (no\_cache=False)
+- Public Methods: 
+    1. read(table: string, key: string, max\_age=None, no\_cache=False): value (JSON)
+    2. write(table: string, op\_list: list): status (string), throws StaleException
+- Internal Objects: 
+    1. HTTP library for establishing/managing connection to server (in Python, urllib3)
+    2. map for maintaining cache
+- Exceptions: 
+    1. StaleException
+    2. StaleException(Value_time: int)
+
+## Cache Structure Description
+
+The cache internally maintains (1) a connection to the DB server and (2) a map of cache entries.  
+
+### Database Connection
+
+The cache should utilize an HTTP Request/Response library (e.g., in Python, urllib3) to establish a connection with the DB server and read and write values through its course of use.  We describe in the HTTP Message Formatting section the expected format for DB requests and responses.
+
+[Here are examples](https://urllib3.readthedocs.org/en/latest/) of using the urllib3 library in Python to open and use a connection.
+
+### Map of Cache Entries
+
+We propose two possible designs for implementing the cache; the preferable of the two designs depends largely on the existing libraries available in the application language.
+
+#### Two-Level Map Implementation
+
+The cache is a two-level map of the following type:
+
+    table : key -> value-time -> value : read-time
+
+In Python, the top-level map is a standard Python dictionary, and the inner map is simply an array of `(value-time, value : read-time)` tuples.
+
+Specifically, the `table : key` keys are composed of the concatenation of the table identifier string and the key string.  Each `table:key` maps to an arraylist with entries of the following type:
+
+    (value-time, value : read-time)   
+
+The cache entries themselves are stored in these arraylists.  The values are tuples derived from the value timestamps and `value : read-time` concatenated strings.  
+
+If the arraylists are unsorted, then `add` is *O(1)*, and `find` (floor), `replace`, and `remove` are each *O(n)*, where `n` is the length of the arraylist.  If the arraylists are sorted (which takes *O(nlogn)*), all the operations are *O(n)*.  Thus, the arraylists do not need to be sorted, as sorting would not improve the asymptotic runtimes of these operations.
+
+Note that the cache can indicate the DB did not have a value for a certain key at a given time by storing None in Python (null in other languages) in the `(value, read-time)` entry. 
+
+#### Skiplist Implementation
+
+If an appropriate skiplist library is available, the cache can be implemented as a skiplist which supports `floor` and `ceiling` operations.  The cache is then a single-level map and has the following type:
```

Comment:
> If an appropriate skiplist library is available, the cache can be implemented as a skiplist which supports `floor` and `ceiling` operations

Wordsmithing: If a skiplist that supports `floor` and `ceiling` operations is available, the cache can be implemented using that.